man/8/qmail-fixq.8
.TH qmail-fixq 8 "" eQmail
.SH NAME
qmail-fixq \- check the queue and fix possible issues of the queue
.SH SYNOPSIS
.B qmail-fixq \fI[ -c | -f | [ -h ]\fR
.SH DESCRIPTION
.B qmail-fixq
goes through the directory structure of the queue and reads splitting and access rights.
\fBqmail-fixq\fR has the following options:
.TP
\fI-c\fR check the queue and print results
.TP
\fI-f\fR fix issues of the queue
.TP
\fI-h\fR print help
.SH "SEE ALSO"
qmail-qstat(8)


CHANGES
commit 5d9fb974fa147f29fe7d8fc438ffe9cfb70d803d
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 15 16:09:15 2017 +0100

    removed subgetopt/sgetopt

commit cabc31d1eae1b9ffc056142afa9f58ad3e1e2ecf
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 15 13:58:54 2017 +0100

    reworked 'configure' script

commit 6f57aeea75aa12e8fca937de3761d0b2bf0c5d07
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 18:10:04 2017 +0100

    qmail-inject: switched to buffer

commit e3930d23f3df952b6b1276cd7e23722c9dac793f
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 16:59:45 2017 +0100

    qmail-pw2u: switched to 'buffer'

commit bdc2690aecb77ba57aabaf4b4c399ff36cef9789
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 16:45:29 2017 +0100

    tcp-env: switched to getoptb (buffer)

commit e9d22dd4fff1b4fe8395da4d41862e78036ee53f
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 15:36:56 2017 +0100

    preline: switched to 'buffer'

commit ff259662a326a975e56aab07fceae08d9e006298
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 11:25:55 2017 +0100

    added 'socket' lib from ucspi-tcp-088

commit 900d1d7b469ccb793b0b254965523c1f7f3c8a8b
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 08:22:24 2017 +0100

    reworked stralloc lib

commit 71e628955516f4055466902a8e1ca100f56f7ba2
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 13 15:40:54 2017 +0100

    improved build/install system, code clean-up

commit 176589f5c769cc60e9321f17ca3cd294a9f067c0
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 10 11:19:09 2017 +0100

    reworked env lib

commit a0b69812ab0e25089f48b21bc69e5dc5597b0a9f
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 9 14:12:38 2017 +0100

    added lib getln_buf

commit 649080aedb1a706827cb1dc935626516cbca9d9e
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 9 14:11:39 2017 +0100

    install target in Makefile

commit 4a591da4d737131fcc9d10458fed0cbfd9af0c17
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 16:31:16 2017 +0100

    added fuction 'scan_xlong' to scan lib

commit ed716f2d7587762ed6b1227700e430825f764e03
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 16:19:04 2017 +0100

    switched to newer sig() lib from ucspi-tcp-0.88

commit d5b1a2055373b39aa59273a74a775a923c120a37
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 16:06:12 2017 +0100

    added tai, taia and iopauce from ucspi-tcp-0.88

commit 205fbd816da6b4098c748e9b293a1337188acc55
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 10:09:59 2017 +0100

    replaced vfork() by fork()

commit 749e25226bf845a98e87217398ec005ed0ccd983
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 09:12:24 2017 +0100

    updated to tls-patch 20160918

commit 6d1a4dbf3da483e0e9c4beab21538bb75ddedee0
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 6 16:49:27 2017 +0100

    switch to new(er) cdb lib

commit e4fa52e96d947e02d76d11106a8e614d91d3f28f
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 6 11:41:55 2017 +0100

    qmail-local: switched to 'getoptb' and 'buffer'

commit 528df12a41c5156c1f05ad83749b0543389a2de5
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 6 09:38:42 2017 +0100

    added 'getoptb'

commit 83e8b06c4bb1aa8dcf1c54f787bea79b92ee0153
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 3 12:18:21 2017 +0100

    added mkrsadhkeys mksrvrcerts

commit 6f739706612ca4d52b6fa4c4fdefe22ed07397c6
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 3 12:06:49 2017 +0100

    re-worked fmtqfn.{h,c}

commit 96552e68348125c1f3c1f401780a67b703f8d422
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 3 11:19:48 2017 +0100

    added 'buffer' lib

commit 19c9c42879ca13d843640dc935f76d3712153c27
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 2 16:48:22 2017 +0100

    added qmail-shcfg, qmail-fixq

commit 914ea8122b3591adf48f825eb7f9a6ab9db6690e
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 2 16:46:48 2017 +0100

    replaced qmail-qstat

commit e5ba1ad6d98f8a9c51c335a220d4c9d0b2a1ad2a
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 2 16:20:33 2017 +0100

    replaced outdated install script after rollback

commit 576e6e365f60931ef2276c9b5b1458843e441269
Merge: 446793b e595a09
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 14:49:36 2016 +0200

    fixed conflicts

commit e595a096efbdd0317e052b57e3c48df8daa51345
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 14:36:37 2016 +0200

    add files before solve conflicts

commit fc00052ccc96e45be2699bf4fb9fcd224a2b828a
Merge: a76e297 416edce
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 14:34:46 2016 +0200

    Merge branch 'master' into dev

commit 446793ba603e3fc97a3c460efe2e027809035a23
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 12:14:51 2016 +0200

    resolve conflicts

commit b651d37a53bde2074cf31020f224658d964e4868
Merge: 360e06b f643417
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 12:08:28 2016 +0200

    resolve conflicts

commit 360e06b890427969757aa403fc82f4fb947cd58a
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 11:35:14 2016 +0200

    1.09 final

commit a76e297b12ebdf5967e264d36691cc1774c635de
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Apr 8 13:11:17 2016 +0200

    1.10p3

commit f64341737d56aa05fca8695e71f658c6589a6d82
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 29 17:00:52 2016 +0200

    Update CHANGLOG-1.09

commit 57b16e354477498fc3f6c443bbbfd6b146b9c99c
Merge: 416edce e4c20f9
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 29 16:59:32 2016 +0200

    Merge pull request #2 from kp-org/dev

    eQmail-1.09 RC2

commit e4c20f947c4251e23bf619a2158e4d552fbe3165
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 29 16:39:57 2016 +0200

    eQmail-1.09 RC2

commit 8a1e087026b1ca66f0bcccff7214833d0bd1aba1
Author: Kai Peter <kp@openqmail.org>
Date:   Sun Mar 27 01:54:09 2016 +0100

    1.10p2

commit cfc7ed9f30bdc01ddc81403c629979ad96562d6f

qmail-fixq.sh
#!/bin/sh

CFGFILE="conf.tmp"
OUTFILE="qmail-fixq"

[ -f "$CFGFILE" ] || exit 1

# the 'stat' command is different on Linux/BSD
STATOPTS="-f %Lp"
cat systype | grep -i linux >/dev/null
if [ $? -eq 0 ] ; then STATOPTS="-c %a" ; fi

cat warn-auto.sh > "$OUTFILE"
echo "" >> "$OUTFILE"
echo $(cat "$CFGFILE" | grep ^HOMEDIR | sed s}HOMEDIR}QMAILDIR}g) >> "$OUTFILE"
echo $(cat "$CFGFILE" | grep ^UIDQ)                               >> "$OUTFILE"
echo $(cat "$CFGFILE" | grep ^UIDR)                               >> "$OUTFILE"
echo $(cat "$CFGFILE" | grep ^UIDS)                               >> "$OUTFILE"
echo $(cat "$CFGFILE" | grep ^GIDQ)                               >> "$OUTFILE"
echo $(cat "$CFGFILE" | grep ^SPLIT)                              >> "$OUTFILE"
echo "" >> "$OUTFILE"

echo "VARDIR=\"\$QMAILDIR/var\""  >> "$OUTFILE"
echo "
showHelp() {
  echo \"Usage: $OUTFILE [option [-m]\"
  echo
  echo -e \"  -c\tcheck structure of the queue for issues (default)\"
  echo -e \"  -f\tfix issues of the queue\"
  echo -e \"  -h\tShow this help\"
#  echo -e \"  -m\tincludes check of messages (this could take a long time)\"  # useless in practice, messages
                                                                               # will be fixed always
  echo
  exit 0
}" >> "$OUTFILE"

echo "fixqueue() {"                                               >> "$OUTFILE"
echo "  echo -n \"Fixing queue ...\""                             >> "$OUTFILE"
# the code comes from the 'install' script
fln=`cat install | grep -n '^# queue' | cut -d: -f1`   # get the first line
lln=`cat install | grep -n '^# bin' | cut -d: -f1`     # lastline
fln=`expr $fln + 1`                                    # skip one line on begin
lines=`expr $lln - $fln - 1`                           # ... and on the end
#cat install | grep -A $l '^# queue' >> $OUTFILE
cat install | tail -n +$fln | head -n $lines | sed 's/^/  /'      >> "$OUTFILE"

echo "  echo \"done!\"" >> "$OUTFILE"
echo "}"                >> "$OUTFILE"
echo "chkqueue() {
  echo \"Checking queue for issues ...\"
  SPLITMESS=\`find \$VARDIR/queue/mess/*   -type d -print | wc -l\`
  SPLITINFO=\`find \$VARDIR/queue/info/*   -type d -print | wc -l\`
  SPLITINTD=\`find \$VARDIR/queue/intd/*   -type d -print | wc -l\`
  SPLITLOCL=\`find \$VARDIR/queue/local/*  -type d -print | wc -l\`
  SPLITREMT=\`find \$VARDIR/queue/remote/* -type d -print | wc -l\`
  SPLITTODO=\`find \$VARDIR/queue/todo/*   -type d -print | wc -l\`
  if [ \"\$SPLIT\" -ne \"\$SPLITMESS\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITINFO\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITINTD\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITLOCL\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITREMT\" ]; then SPLITERR=1 ; fi
  if [ \"\$SPLIT\" -ne \"\$SPLITTODO\" ]; then SPLITERR=1 ; fi

  if [ \"\$SPLITERR\" ] ; then
      echo \"invalid queue split\" ; ERR=1
   else
      echo \"splitting ok!\"
  fi
  # check the directory structure
  # group have to be \'qmail\' always (recursive)
  GRPNAME=\`getent group \$GIDQ | awk -F: '{print \$1}'\`
  find \$VARDIR/queue/* -type d -print | xargs ls -ld | awk '{print \$4}' | grep -v \$GRPNAME >/dev/null
  if [ \$? -eq 0 ] ; then GRPERR=1 ; fi

  # check for uid qmailq/qmails on splitted directories
  USRNAME=\`id -un \$UIDQ\`
  for f in intd mess todo local remote info
  do
    if [ \"\$f\" = \"local\" ] ; then USRNAME=\`id -un \$UIDS\` ; fi

#echo \$f
    ls -ld \$VARDIR/queue/\$f | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
    if [ \$? -eq 0 ] ; then USRERR=1 ; fi
#find \$VARDIR/queue/\$f/*
    find \$VARDIR/queue/\$f/* | xargs ls -ld | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
    if [ \$? -eq 0 ] ; then USRERR=1 ; fi
  done
  # check pid, bounce
  ls -ld \$VARDIR/queue/bounce | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
    if [ \$? -eq 0 ] ; then USRERR=1 ; fi
  USRNAME=\`id -un \$UIDQ\`   # until now USRNAME was 'qmails'
  ls -ld \$VARDIR/queue/pid    | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
    if [ \$? -eq 0 ] ; then USRERR=1 ; fi

  # special: lock
#echo U: \$USRNAME
  ls -ld \$VARDIR/queue/lock   | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
  if [ \$? -eq 0 ] ; then USRERR=1 ; fi
  USRNAME=\`id -un \$UIDS\`   # ... back to USRNAME was 'qmails'
  ls -l \$VARDIR/queue/lock/trigger   | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
  if [ \$? -eq 0 ] ; then USRERR=1 ; fi
#echo U: \$USRNAME
  ls -l \$VARDIR/queue/lock/sendmutex | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
  if [ \$? -eq 0 ] ; then USRERR=1 ; fi

  USRNAME=\`id -un \$UIDR\`   # ... and now we need USRNAME 'qmailr'
#echo U: \$USRNAME
  ls -l \$VARDIR/queue/lock/tcpto | awk '{print \$3}' | grep -v \$USRNAME >/dev/null
  if [ \$? -eq 0 ] ; then USRERR=1 ; fi

  if [ \"\$USRERR\" ] || [ \"\$GRPERR\" ] ; then
     echo \"invalid owner(s) or group!\"
   else
     echo \"user/group ok!\"
  fi

  # check permissions w/ and w/o subdirectories
  stat $STATOPTS \`ls -d \$VARDIR/queue/mess\`     | grep -v 750 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/mess/*\`   | grep -v 750 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/todo\`     | grep -v 750 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/todo/*\`   | grep -v 750 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/info\`     | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/info/*\`   | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/intd\`     | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/intd/*\`   | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/local\`    | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/local/*\`  | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/remote\`   | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi
  stat $STATOPTS \`ls -d \$VARDIR/queue/remote/*\` | grep -v 700 >/dev/null
  if [ \$? -eq 0 ] ; then PRMERR=1 ; fi

  if [ \"\$PRMERR\" ] ; then
     echo \"invalid permissions(s)!\"
   else
     echo \"permissions ok!\"
  fi

#  echo
}" >> "$OUTFILE"

echo "
if [ ! \"\$1\" ] ; then chkqueue ; exit 0 ; fi
while getopts ":hcf" o; do
    case \"\${o}\" in
        h) showHelp;;
        f) fixqueue;;
        c) chkqueue;;
        m) FIXMSG=\"Y\" ; fixqueue;;
        *) echo "Invalid option!" ; showHelp;;
    esac
done ; shift $((OPTIND-1))

exit 0" >> "$OUTFILE"

chmod 755 $OUTFILE
